src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -299,3 +350,72 @@ SDCPUThread::startup()
     _committedState->startup();
     // TODO other things?
 }
+
+Fault
+SDCPUThread::MemIface::initiateMemRead(InflightInst& ref, Addr addr,
+                                       unsigned int size, Request::Flags flags)
+{
+    const int asid = 0;
+    const Addr pc = 0; // TODO extract this from inflightinst, but otherwise
+                       //      only used for tracing/debugging.
+    const MasterID mid = sdCPUThread._cpuPtr->dataMasterId();
+    const BaseTLB::Mode mode = BaseTLB::Read;
+    const ThreadContext* tc = sdCPUThread.getThreadContext();
+    const ContextID cid = tc->contextId();
+
+    RequestPtr req = make_shared<Request>(asid, addr, size, flags, mid, pc,
+                                          cid);
+
+    // For tracing?
+    req->taskId(sdCPUThread._cpuPtr->taskId());
+
+    const unsigned cache_line_size = sdCPUThread._cpuPtr->cacheLineSize();
+    const Addr split_addr = roundDown(addr + size - 1, cache_line_size);
+
+    if (split_addr > addr) { // The request spans two cache lines
+        panic("SDCPUThread currently rejects split requests.");
+        // TODO Deal with it and send two requests to the CPU.
+    }
+
+    sdCPUThread._cpuPtr->requestDataAddrTranslation(req, tc,
+        [&sdCPUThread, &ref](Fault fault, const RequestPtr& req) {
+            sdCPUThread.onDataAddrTranslated(fault, req, ref, false);
+        });
+
+    return NoFault;
+}
+
+Fault
+SDCPUThread::MemIface::writeMem(InflightInst& ref, uint8_t *data,
```

Comment:
There's a lot of duplicated code between this function and `initiateMemRead()`. Any way to factor some out into its own function?

Reply:
I had that thought too, but I'm also not quite done with the writeMem function yet, since I need to find a good way to pass the uint8_t* data param through to when the actual packet is being sent. Once that is sorted out, I'll think about consolidating the code.